# PHASE 2: DATA UNDERSTANDING
The objective of this phase is to load the dataset, understand its structure, and perform initial exploration.

A. Loading and Inspecting the data

In [1]:
import os
import pandas as pd

#  Set root data folder
data_root = "./data"

#  Recursively collect all .csv files
csv_files = []
for root, dirs, files in os.walk(data_root):
    for file in files:
        if file.lower().endswith(".csv"):
            csv_files.append(os.path.join(root, file))

print(f" Total CSV files found: {len(csv_files)}")

# Try loading the first file to inspect
first_file = csv_files[0]
print(f" Previewing: {first_file}")

try:
    df = pd.read_csv(first_file, encoding="utf-8")  # adjust encoding if needed
    print(" Loaded as CSV")
except Exception as e_csv:
    try:
        df = pd.read_excel(first_file)
        print("Loaded as Excel")
    except Exception as e_excel:
        print("Failed to load file as CSV or Excel.")
        print("CSV Error:", e_csv)
        print("Excel Error:", e_excel)
        df = None

#  Preview the structure
if df is not None:
    print("🔹 Shape:", df.shape)
    print("🔹 Columns:", df.columns.tolist())
    display(df.head())


 Total CSV files found: 60
 Previewing: ./data\2020\April2020TransBorderRawData\Apr 2020\dot1_0420.csv
 Loaded as CSV
🔹 Shape: (24384, 14)
🔹 Columns: ['TRDTYPE', 'USASTATE', 'DEPE', 'DISAGMOT', 'MEXSTATE', 'CANPROV', 'COUNTRY', 'VALUE', 'SHIPWT', 'FREIGHT_CHARGES', 'DF', 'CONTCODE', 'MONTH', 'YEAR']


TRDTYPE USASTATE  DEPE  DISAGMOT MEXSTATE CANPROV  COUNTRY    VALUE  \
0        1       AK  0115         5      NaN      XB     1220     4660   
1        1       AK  0901         5      NaN      XO     1220    14360   
2        1       AK  20XX         1       XX     NaN     2010  4293733   
3        1       AK  20XX         3      NaN      XA     1220    28283   
4        1       AK  20XX         3      NaN      XA     1220    29848   

     SHIPWT  FREIGHT_CHARGES   DF CONTCODE  MONTH  YEAR  
0         0               67  2.0        X      4  2020  
1         0              282  1.0        X      4  2020  
2  24971000                0  1.0        0      4  2020  
3       443              563  1.0        X      4  2020  
4        69              538  2.0        X      4  2020

In [ ]:
import os
import pandas as pd

#  Set root data folder
data_root = "./data"

#  Recursively collect all .csv files
csv_files = []
for root, dirs, files in os.walk(data_root):
    for file in files:
        if file.lower().endswith(".csv"):
            csv_files.append(os.path.join(root, file))

print(f" Total CSV files found: {len(csv_files)}")

# Try loading the first file to inspect
first_file = csv_files[0]
print(f" Previewing: {first_file}")

try:
    df = pd.read_csv(first_file, encoding="utf-8")  # adjust encoding if needed
    print(" Loaded as CSV")
except Exception as e_csv:
    try:
        df = pd.read_excel(first_file)
        print("Loaded as Excel")
    except Exception as e_excel:
        print("Failed to load file as CSV or Excel.")
        print("CSV Error:", e_csv)
        print("Excel Error:", e_excel)
        df = None

#  Preview the structure
if df is not None:
    print("🔹 Shape:", df.shape)
    print("🔹 Columns:", df.columns.tolist())
    display(df.head())


 Total CSV files found: 60
 Previewing: ./data\2020\April2020TransBorderRawData\Apr 2020\dot1_0420.csv
 Loaded as CSV
🔹 Shape: (24384, 14)
🔹 Columns: ['TRDTYPE', 'USASTATE', 'DEPE', 'DISAGMOT', 'MEXSTATE', 'CANPROV', 'COUNTRY', 'VALUE', 'SHIPWT', 'FREIGHT_CHARGES', 'DF', 'CONTCODE', 'MONTH', 'YEAR']


TRDTYPE USASTATE  DEPE  DISAGMOT MEXSTATE CANPROV  COUNTRY    VALUE  \
0        1       AK  0115         5      NaN      XB     1220     4660   
1        1       AK  0901         5      NaN      XO     1220    14360   
2        1       AK  20XX         1       XX     NaN     2010  4293733   
3        1       AK  20XX         3      NaN      XA     1220    28283   
4        1       AK  20XX         3      NaN      XA     1220    29848   

     SHIPWT  FREIGHT_CHARGES   DF CONTCODE  MONTH  YEAR  
0         0               67  2.0        X      4  2020  
1         0              282  1.0        X      4  2020  
2  24971000                0  1.0        0      4  2020  
3       443              563  1.0        X      4  2020  
4        69              538  2.0        X      4  2020

Before merging, I'm going to validate column structure across all 60 files to ensure that there are no missing or extra columns, column order is consistent so that I know early if any files need fixing:



In [2]:

def check_column_structures(file_list):
    structure_map = {}

    for file in file_list:
        try:
            df = pd.read_csv(file, nrows=5)  # Only read a few rows for speed
            col_set = tuple(df.columns)  # Keep column order too
            structure_map.setdefault(col_set, []).append(file)
        except Exception as e:
            print(f"Failed to read {file}: {e}")

    print(f"\n Found {len(structure_map)} unique column structures:\n")

    for i, (col_set, files) in enumerate(structure_map.items(), 1):
        print(f"Structure {i}: {len(files)} file(s)")
        print("Columns:", col_set)
        print("Sample file:", files[0])
        print("-" * 50)

    return structure_map
structure_report = check_column_structures(csv_files)



Failed to read ./data\2020\July2020TransBorderRawData\__MACOSX\._dot1_0720.csv: 'utf-8' codec can't decode byte 0xb1 in position 45: invalid start byte
Failed to read ./data\2020\July2020TransBorderRawData\__MACOSX\._dot1_ytd_0720.csv: 'utf-8' codec can't decode byte 0xb1 in position 45: invalid start byte
Failed to read ./data\2020\July2020TransBorderRawData\__MACOSX\._dot2_0720.csv: 'utf-8' codec can't decode byte 0xb1 in position 45: invalid start byte
Failed to read ./data\2020\July2020TransBorderRawData\__MACOSX\._dot2_ytd_0720.csv: 'utf-8' codec can't decode byte 0xb1 in position 45: invalid start byte
Failed to read ./data\2020\July2020TransBorderRawData\__MACOSX\._dot3_0720.csv: 'utf-8' codec can't decode byte 0xb1 in position 45: invalid start byte
Failed to read ./data\2020\July2020TransBorderRawData\__MACOSX\._dot3_ytd_0720.csv: 'utf-8' codec can't decode byte 0xb1 in position 45: invalid start byte

 Found 3 unique column structures:

Structure 1: 18 file(s)
Columns: ('TRDT

This output shows that there are 3 different/unique columns structure so I'm going to merge each structure separately and create 3 master DataFrames:

- Function to load and detect a file

In [3]:
import pandas as pd

def safe_load_csv(filepath, columns_expected):
    """Load CSV and check if it matches expected structure"""
    try:
        df = pd.read_csv(filepath)
        if set(columns_expected).issubset(set(df.columns)):
            df["source_file"] = filepath
            return df
        else:
            print(f" Skipped {filepath} (unexpected columns)")
            return None
    except Exception as e:
        print(f" Failed to load {filepath}: {e}")
        return None


- Function for dot1 (State-level freight totals -no commodity code) loader

In [4]:
def load_dot1_files(file_list):
    dot1_columns = [
        'TRDTYPE', 'USASTATE', 'DEPE', 'DISAGMOT', 'MEXSTATE', 'CANPROV',
        'COUNTRY', 'VALUE', 'SHIPWT', 'FREIGHT_CHARGES', 'DF', 'CONTCODE',
        'MONTH', 'YEAR'
    ]def load_dot1_files(file_list):
    dot1_columns = [
        'TRDTYPE', 'USASTATE', 'DEPE', 'DISAGMOT', 'MEXSTATE', 'CANPROV',
        'COUNTRY', 'VALUE', 'SHIPWT', 'FREIGHT_CHARGES', 'DF', 'CONTCODE',
        'MONTH', 'YEAR'
    ]
    dot1_data = []
    for file in file_list:
        df = safe_load_csv(file, dot1_columns)
        if df is not None and len(df.columns) == len(dot1_columns):
            df["structure"] = "dot1"
            dot1_data.append(df)
    return pd.concat(dot1_data, ignore_index=True) if dot1_data else pd.DataFrame()

    dot1_data = []
    for file in file_list:
        df = safe_load_csv(file, dot1_columns)
        if df is not None and len(df.columns) == len(dot1_columns):
            df["structure"] = "dot1"
            dot1_data.append(df)
    return pd.concat(dot1_data, ignore_index=True) if dot1_data else pd.DataFrame()


- Function for dot2 (Commodity-by-state -adds COMMODITY2) loader

In [5]:
def load_dot2_files(file_list):
    dot2_columns = [
        'TRDTYPE', 'USASTATE', 'COMMODITY2', 'DISAGMOT', 'MEXSTATE', 'CANPROV',
        'COUNTRY', 'VALUE', 'SHIPWT', 'FREIGHT_CHARGES', 'DF', 'CONTCODE',
        'MONTH', 'YEAR'
    ]
    dot2_data = []
    for file in file_list:
        df = safe_load_csv(file, dot2_columns)
        if df is not None and len(df.columns) == len(dot2_columns):
            df["structure"] = "dot2"
            dot2_data.append(df)
    return pd.concat(dot2_data, ignore_index=True) if dot2_data else pd.DataFrame()


- Function for dot3 (National-level commodity totals-no USASTATE) loader

In [6]:
def load_dot3_files(file_list):
    dot3_columns = [
        'TRDTYPE', 'DEPE', 'COMMODITY2', 'DISAGMOT', 'COUNTRY', 'VALUE',
        'SHIPWT', 'FREIGHT_CHARGES', 'DF', 'CONTCODE', 'MONTH', 'YEAR'
    ]
    dot3_data = []
    for file in file_list:
        df = safe_load_csv(file, dot3_columns)
        if df is not None and len(df.columns) == len(dot3_columns):
            df["structure"] = "dot3"
            dot3_data.append(df)
    return pd.concat(dot3_data, ignore_index=True) if dot3_data else pd.DataFrame()


Now loading each of the 3 dataset.....

In [7]:

# ---------- 1. Safe loader function ----------
def safe_load_csv(filepath, expected_len):
    try:
        df = pd.read_csv(filepath, dtype=str, low_memory=False)
        actual_cols = df.columns.tolist()
        print(f"\n {os.path.basename(filepath)} columns: {actual_cols}")
        if len(actual_cols) == expected_len:
            df["source_file"] = filepath
            return df
        else:
            print(f" Skipped {filepath} (expected {expected_len} cols, got {len(actual_cols)})")
            return None
    except Exception as e:
        print(f" Failed to load {filepath}: {e}")
        return None

# ---------- 2. dot1 ----------
def load_dot1_files(file_list):
    dot1_expected_len = 14
    return pd.concat(
        [df for df in (safe_load_csv(f, dot1_expected_len) for f in file_list) if df is not None],
        ignore_index=True
    ) if file_list else pd.DataFrame()

# ---------- 3. dot2 ----------
def load_dot2_files(file_list):
    dot2_expected_len = 14
    return pd.concat(
        [df for df in (safe_load_csv(f, dot2_expected_len) for f in file_list) if df is not None],
        ignore_index=True
    ) if file_list else pd.DataFrame()

# ---------- 4. dot3 ----------
def load_dot3_files(file_list):
    dot3_expected_len = 12
    return pd.concat(
        [df for df in (safe_load_csv(f, dot3_expected_len) for f in file_list) if df is not None],
        ignore_index=True
    ) if file_list else pd.DataFrame()

# ---------- 5. Filter file list ----------
dot1_files = [f for f in csv_files if "dot1" in f.lower() and "__macosx" not in f.lower()]
dot2_files = [f for f in csv_files if "dot2" in f.lower() and "__macosx" not in f.lower()]
dot3_files = [f for f in csv_files if "dot3" in f.lower() and "__macosx" not in f.lower()]

# ---------- 6. Load all ----------
df_dot1 = load_dot1_files(dot1_files)
df_dot2 = load_dot2_files(dot2_files)
df_dot3 = load_dot3_files(dot3_files)

# ---------- 7. Preview results ----------
print(" dot1:", df_dot1.shape)
print(" dot2:", df_dot2.shape)
print(" dot3:", df_dot3.shape)



 dot1_0420.csv columns: ['TRDTYPE', 'USASTATE', 'DEPE', 'DISAGMOT', 'MEXSTATE', 'CANPROV', 'COUNTRY', 'VALUE', 'SHIPWT', 'FREIGHT_CHARGES', 'DF', 'CONTCODE', 'MONTH', 'YEAR']

 dot1_ytd_0420.csv columns: ['TRDTYPE', 'USASTATE', 'DEPE', 'DISAGMOT', 'MEXSTATE', 'CANPROV', 'COUNTRY', 'VALUE', 'SHIPWT', 'FREIGHT_CHARGES', 'DF', 'CONTCODE', 'MONTH', 'YEAR']

 dot1_0820.csv columns: ['TRDTYPE', 'USASTATE', 'DEPE', 'DISAGMOT', 'MEXSTATE', 'CANPROV', 'COUNTRY', 'VALUE', 'SHIPWT', 'FREIGHT_CHARGES', 'DF', 'CONTCODE', 'MONTH', 'YEAR']

 dot1_ytd_0820.csv columns: ['TRDTYPE', 'USASTATE', 'DEPE', 'DISAGMOT', 'MEXSTATE', 'CANPROV', 'COUNTRY', 'VALUE', 'SHIPWT', 'FREIGHT_CHARGES', 'DF', 'CONTCODE', 'MONTH', 'YEAR']

 dot1_0220.csv columns: ['TRDTYPE', 'USASTATE', 'DEPE', 'DISAGMOT', 'MEXSTATE', 'CANPROV', 'COUNTRY', 'VALUE', 'SHIPWT', 'FREIGHT_CHARGES', 'DF', 'CONTCODE', 'MONTH', 'YEAR']

 dot1_ytd_0220.csv columns: ['TRDTYPE', 'USASTATE', 'DEPE', 'DISAGMOT', 'MEXSTATE', 'CANPROV', 'COUNTRY', 'VALU

In [8]:
print("dot1 files:", len(dot1_files))
print("dot2 files:", len(dot2_files))
print("dot3 files:", len(dot3_files))
print("Sample dot1 file:", dot1_files[0] if dot1_files else "None")


dot1 files: 18
dot2 files: 18
dot3 files: 18
Sample dot1 file: ./data\2020\April2020TransBorderRawData\Apr 2020\dot1_0420.csv


In [9]:
import pandas as pd
import os
from glob import glob

# Define root folder
root_dir = './data/2020'

# Collect files per category
dot1_files = sorted(glob(os.path.join(root_dir, '**', 'dot1*.csv'), recursive=True))
dot2_files = sorted(glob(os.path.join(root_dir, '**', 'dot2*.csv'), recursive=True))
dot3_files = sorted(glob(os.path.join(root_dir, '**', 'dot3*.csv'), recursive=True))

# Load and combine all dot1, dot2, dot3 files into separate DataFrames
def load_and_combine(files):
    dfs = []
    for file in files:
        try:
            df = pd.read_csv(file, low_memory=False, encoding='latin1')  # Avoid utf-8 errors
            dfs.append(df)
        except Exception as e:
            print(f" Failed to load {file}: {e}")
    return pd.concat(dfs, ignore_index=True)

dot1 = load_and_combine(dot1_files)
dot2 = load_and_combine(dot2_files)
dot3 = load_and_combine(dot3_files)

print(f" dot1 shape: {dot1.shape}")
print(f" dot2 shape: {dot2.shape}")
print(f" dot3 shape: {dot3.shape}")

# View first 5 rows of each merged dataset
print(" dot1 (first 5 rows):")
display(dot1.head())

print("\n dot2 (first 5 rows):")
display(dot2.head())

print("\n dot3 (first 5 rows):")
display(dot3.head())



 dot1 shape: (1393769, 14)
 dot2 shape: (3818938, 14)
 dot3 shape: (892060, 12)
 dot1 (first 5 rows):


TRDTYPE USASTATE  DEPE  DISAGMOT MEXSTATE CANPROV  COUNTRY    VALUE  \
0        1       AK  0115         5      NaN      XB     1220     4660   
1        1       AK  0901         5      NaN      XO     1220    14360   
2        1       AK  20XX         1       XX     NaN     2010  4293733   
3        1       AK  20XX         3      NaN      XA     1220    28283   
4        1       AK  20XX         3      NaN      XA     1220    29848   

     SHIPWT  FREIGHT_CHARGES   DF CONTCODE  MONTH  YEAR  
0         0               67  2.0        X      4  2020  
1         0              282  1.0        X      4  2020  
2  24971000                0  1.0        0      4  2020  
3       443              563  1.0        X      4  2020  
4        69              538  2.0        X      4  2020


 dot2 (first 5 rows):


TRDTYPE USASTATE  COMMODITY2  DISAGMOT MEXSTATE CANPROV  COUNTRY   VALUE  \
0        1       AK           2         5       NL     NaN     2010   22370   
1        1       AK           3         1      NaN      XB     1220  108133   
2        1       AK           3         1      NaN      XC     1220  809104   
3        1       AK           3         5      NaN      XC     1220  887888   
4        1       AK           3         5      NaN      XO     1220   76006   

   SHIPWT  FREIGHT_CHARGES   DF CONTCODE  MONTH  YEAR  
0       0                0  1.0        0      4  2020  
1   24698             1482  1.0        X      4  2020  
2   99790            15864  1.0        X      4  2020  
3       0            17546  1.0        X      4  2020  
4       0             1665  1.0        X      4  2020


 dot3 (first 5 rows):


TRDTYPE  DEPE  COMMODITY2  DISAGMOT  COUNTRY  VALUE  SHIPWT  \
0        1  0101           9         5     1220   2864       0   
1        1  0101          19         5     1220  12690       0   
2        1  0101          20         5     1220   6699       0   
3        1  0101          25         6     1220  25854       0   
4        1  0101          28         5     1220  16007       0   

   FREIGHT_CHARGES   DF CONTCODE  MONTH  YEAR  
0               56  2.0        X      4  2020  
1              241  1.0        X      4  2020  
2              131  1.0        X      4  2020  
3              506  1.0        X      4  2020  
4              313  2.0        X      4  2020

In summary, I have done these in the first phase :

1. File Organization & Structure Check
- Unzipped and placed all CSV files from subfolders into the `./data` directory.
- A total of **60 CSV files** were detected.

2. Structure Consistency Check
- Validated that files fall into 3 unique structures based on column names:
  - **dot1** files (18): Include `USASTATE`, `DEPE`, `MEXSTATE`, `CANPROV`
  - **dot2** files (18): Similar to dot1 but with `COMMODITY2` replacing `DEPE`
  - **dot3** files (18): Simpler structure with fewer columns; only `DEPE` and `COMMODITY2`

3. Ignored Mac OS Metadata Files
- Skipped unreadable `__MACOSX` files and those with encoding issues.

4. File Classification
- Grouped files into:
  - `dot1_files`: 18 files
  - `dot2_files`: 18 files
  - `dot3_files`: 18 files

5. Structure Validation & Sample Checks
- Ensured all files within each group have consistent column formats.
- Validated sample file headers to match expected schemas.

6. Data Loading Completed
- Successfully loaded and concatenated:
  - `dot1`:  Shape = (1,393,769, 15)
  - `dot2`:  Shape = (3,818,938, 15)
  - `dot3`:  Shape = (892,060, 13)
- Resulting in a total of **6,104,767 rows** across 54 clean files.

---

⏭️ **Next Phase:** Data Cleaning & Wrangling


# PHASE 3: DATA CLEANING & WRANGLING

The objective of this phase is to prepare the data for analysis by ensuring it is clean, consistent, and structured appropriately. 
By the end of this phase, our data should be **clean**, **uniform**, and **ready** for analysis.

a. Checking for Missing or Null Values in each of the three datasets: dot1, dot2, and dot3.....


In [10]:

import pandas as pd

# Compute missing values
dot1_missing = dot1.isnull().sum()
dot2_missing = dot2.isnull().sum()
dot3_missing = dot3.isnull().sum()

# Combine into one DataFrame
missing_df = pd.DataFrame({
    'dot1_missing': dot1_missing,
    'dot2_missing': dot2_missing,
    'dot3_missing': dot3_missing
}).fillna('—')

# Format numbers with commas
for col in missing_df.columns:
    missing_df[col] = missing_df[col].map(lambda x: f"{int(x):,}" if isinstance(x, (int, float)) and x != '—' else x)

# Display
display(missing_df)


dot1_missing dot2_missing dot3_missing
CANPROV              506,264    1,509,201            —
COMMODITY2                 —            0            0
CONTCODE                   0            0            0
COUNTRY                    0            0            0
DEPE                       0            —            0
DF                   530,290    1,176,621      360,277
DISAGMOT                   0            0            0
FREIGHT_CHARGES            0            0            0
MEXSTATE             993,246    2,544,119            —
MONTH                      0            0            0
SHIPWT                     0            0            0
TRDTYPE                    0            0            0
USASTATE                   0            0            —
VALUE                      0            0            0
YEAR                       0            0            0

dot1 and dot2 both have missing values in MEXSTATE, CANPROV, and DF.

dot3 doesn’t have MEXSTATE, CANPROV, or USASTATE at all, as expected based on its structure.

COMMODITY2 is only relevant in dot2 and dot3, not in dot1.

Most of the core numerical and categorical variables like VALUE, SHIPWT, TRDTYPE, DISAGMOT, and FREIGHT_CHARGES are complete in all datasets.

However, based on the business goals, I will:
 - Fill missing values with "Unknown" for MEXSTATE, CANPROV and DEPE (categorical columns).
 - Fill missing values with the column mean, for DF (numerical columns).

In [11]:
# Columns to fill with 'Unknown'
categorical_missing = ['MEXSTATE', 'CANPROV', 'DEPE']

# Column to fill with mean
numerical_missing = ['DF']

def fill_missing_values(df, cat_cols, num_cols):
    df = df.copy()
    for col in cat_cols:
        if col in df.columns:
            df[col] = df[col].fillna('Unknown')
    for col in num_cols:
        if col in df.columns:
            df[col] = df[col].fillna(df[col].mean())
    return df

# Apply to each dataset
dot1 = fill_missing_values(dot1, categorical_missing, numerical_missing)
dot2 = fill_missing_values(dot2, categorical_missing, numerical_missing)
dot3 = fill_missing_values(dot3, categorical_missing, numerical_missing)

# Summary
print("Missing values handled.")
print(f"dot1: {dot1.isnull().sum().sum()} missing values left")
print(f"dot2: {dot2.isnull().sum().sum()} missing values left")
print(f"dot3: {dot3.isnull().sum().sum()} missing values left")


Missing values handled.
dot1: 0 missing values left
dot2: 0 missing values left
dot3: 0 missing values left


b.  Ensuring correct Data Types for each column


In [12]:
# Function to display data types in a readable format
def display_column_dtypes(df, name):
    print(f"Column data types for {name}:\n")
    print(df.dtypes)
    print("-" * 40)

# Display for all three datasets
display_column_dtypes(dot1, "dot1")
display_column_dtypes(dot2, "dot2")
display_column_dtypes(dot3, "dot3")


Column data types for dot1:

TRDTYPE              int64
USASTATE            object
DEPE                object
DISAGMOT             int64
MEXSTATE            object
CANPROV             object
COUNTRY              int64
VALUE                int64
SHIPWT               int64
FREIGHT_CHARGES      int64
DF                 float64
CONTCODE            object
MONTH                int64
YEAR                 int64
dtype: object
----------------------------------------
Column data types for dot2:

TRDTYPE              int64
USASTATE            object
COMMODITY2           int64
DISAGMOT             int64
MEXSTATE            object
CANPROV             object
COUNTRY              int64
VALUE                int64
SHIPWT               int64
FREIGHT_CHARGES      int64
DF                 float64
CONTCODE            object
MONTH                int64
YEAR                 int64
dtype: object
----------------------------------------
Column data types for dot3:

TRDTYPE              int64
DEPE                

- Actions to be taken based on the output above:
Update the data types for each column across the dot1, dot2, and dot3 datasets based on their semantic meaning:

*1. Categorical Columns (converted to category for efficiency)*:

TRDTYPE, USASTATE, DEPE, DISAGMOT, MEXSTATE, CANPROV, COUNTRY, CONTCODE, COMMODITY2 (where present)

*2. Numeric Columns (converted to appropriate numeric types for analysis)*:

VALUE, SHIPWT, FREIGHT_CHARGES, DF - float64

MONTH, YEAR - int64


In [13]:
# Defining the type conversion mapping:
conversion_map = {
    'TRDTYPE': 'category',
    'USASTATE': 'category',
    'DEPE': 'category',
    'DISAGMOT': 'category',
    'MEXSTATE': 'category',
    'CANPROV': 'category',
    'COUNTRY': 'category',
    'VALUE': 'float64',
    'SHIPWT': 'float64',
    'FREIGHT_CHARGES': 'float64',
    'DF': 'float64',
    'CONTCODE': 'category',
    'MONTH': 'int64',
    'YEAR': 'int64',
    'COMMODITY2': 'category'
}
# Applying the conversion to each dataset
def apply_conversions(df, name):
    for col, dtype in conversion_map.items():
        if col in df.columns:
            df[col] = df[col].astype(dtype)
    print(f" Converted types for {name}")

apply_conversions(dot1, "dot1")
apply_conversions(dot2, "dot2")
apply_conversions(dot3, "dot3")


 Converted types for dot1
 Converted types for dot2
 Converted types for dot3


In [ ]:
# Confirming datatype conversions 


print(" Data types for dot1:")
print(dot1.dtypes)
print("-" * 40)

print(" Data types for dot2:")
print(dot2.dtypes)
print("-" * 40)

print(" Data types for dot3:")
print(dot3.dtypes)
print("-" * 40)
  

 Data types for dot1:
TRDTYPE            category
USASTATE           category
DEPE               category
DISAGMOT           category
MEXSTATE           category
CANPROV            category
COUNTRY            category
VALUE               float64
SHIPWT              float64
FREIGHT_CHARGES     float64
DF                  float64
CONTCODE           category
MONTH                 int64
YEAR                  int64
dtype: object
----------------------------------------
 Data types for dot2:
TRDTYPE            category
USASTATE           category
COMMODITY2         category
DISAGMOT           category
MEXSTATE           category
CANPROV            category
COUNTRY            category
VALUE               float64
SHIPWT              float64
FREIGHT_CHARGES     float64
DF                  float64
CONTCODE           category
MONTH                 int64
YEAR                  int64
dtype: object
----------------------------------------
 Data types for dot3:
TRDTYPE            category
DEPE        

c. Checking for duplicates

In [15]:
# Check for duplicates in dot1
duplicates_dot1 = dot1.duplicated()
print(" Duplicate rows in dot1:", duplicates_dot1.sum())

# Check for duplicates in dot2
duplicates_dot2 = dot2.duplicated()
print(" Duplicate rows in dot2:", duplicates_dot2.sum())

# Check for duplicates in dot3
duplicates_dot3 = dot3.duplicated()
print(" Duplicate rows in dot3:", duplicates_dot3.sum())


 Duplicate rows in dot1: 1161740
 Duplicate rows in dot2: 3184057
 Duplicate rows in dot3: 743538


In [16]:
# Display first few duplicates from each

import pandas as pd
import glob

# Define base paths to the files again
dot1_files = glob.glob("./data/2020/**/*dot1_*.csv", recursive=True)
dot2_files = glob.glob("./data/2020/**/*dot2_*.csv", recursive=True)
dot3_files = glob.glob("./data/2020/**/*dot3_*.csv", recursive=True)

# Load each set of files
def load_and_concat(files):
    dfs = []
    for file in files:
        try:
            df = pd.read_csv(file, encoding='utf-8', low_memory=False)
            dfs.append(df)
        except Exception as e:
            print(f"Failed to read {file}: {e}")
    return pd.concat(dfs, ignore_index=True)

dot1 = load_and_concat(dot1_files)
dot2 = load_and_concat(dot2_files)
dot3 = load_and_concat(dot3_files)

# Check and display duplicates
dot1_duplicates = dot1[dot1.duplicated()]
dot2_duplicates = dot2[dot2.duplicated()]
dot3_duplicates = dot3[dot3.duplicated()]

# Display first few duplicates from each
print("🔍 dot1 duplicate rows:")
display(dot1[dot1.duplicated(keep=False)].sort_values(by=dot1.columns.tolist()).head(10))

print("\n🔍 dot2 duplicate rows:")
display(dot2[dot2.duplicated(keep=False)].sort_values(by=dot2.columns.tolist()).head(10))

print("\n🔍 dot3 duplicate rows:")
display(dot3[dot3.duplicated(keep=False)].sort_values(by=dot3.columns.tolist()).head(10))



🔍 dot1 duplicate rows:


TRDTYPE USASTATE  DEPE  DISAGMOT MEXSTATE CANPROV  COUNTRY  VALUE  \
128479         1       AK  0106         5      NaN      XB     1220  37700   
334120         1       AK  0106         5      NaN      XB     1220  37700   
1341431        1       AK  0106         5      NaN      XB     1220  37700   
1135352        1       AK  0115         5      NaN      XB     1220   3665   
1367381        1       AK  0115         5      NaN      XB     1220   3665   
0              1       AK  0115         5      NaN      XB     1220   4660   
104095         1       AK  0115         5      NaN      XB     1220   4660   
234140         1       AK  0115         5      NaN      XB     1220   4660   
596927         1       AK  0115         5      NaN      XB     1220   4660   
802042         1       AK  0115         5      NaN      XB     1220   4660   

         SHIPWT  FREIGHT_CHARGES   DF CONTCODE  MONTH  YEAR  
128479        0              113  1.0        X      8  2020  
334120        0              113  1.0        X      8  2020  
1341431       0              113  1.0        X      8  2020  
1135352       0              399  1.0        X      9  2020  
1367381       0              399  1.0        X      9  2020  
0             0               67  2.0        X      4  2020  
104095        0               67  2.0        X      4  2020  
234140        0               67  2.0        X      4  2020  
596927        0               67  2.0        X      4  2020  
802042        0               67  2.0        X      4  2020


🔍 dot2 duplicate rows:


TRDTYPE USASTATE  COMMODITY2  DISAGMOT MEXSTATE CANPROV  COUNTRY  \
138095         1       AK           2         5       BC     NaN     2010   
495440         1       AK           2         5       BC     NaN     2010   
984375         1       AK           2         5       BC     NaN     2010   
1130618        1       AK           2         5       BC     NaN     2010   
1494110        1       AK           2         5       BC     NaN     2010   
2053964        1       AK           2         5       BC     NaN     2010   
2548399        1       AK           2         5       BC     NaN     2010   
2834024        1       AK           2         5       BC     NaN     2010   
3257268        1       AK           2         5       BC     NaN     2010   
64884          1       AK           2         5       BC     NaN     2010   

         VALUE  SHIPWT  FREIGHT_CHARGES   DF CONTCODE  MONTH  YEAR  
138095   14400       0                0  1.0        0      2  2020  
495440   14400       0                0  1.0        0      2  2020  
984375   14400       0                0  1.0        0      2  2020  
1130618  14400       0                0  1.0        0      2  2020  
1494110  14400       0                0  1.0        0      2  2020  
2053964  14400       0                0  1.0        0      2  2020  
2548399  14400       0                0  1.0        0      2  2020  
2834024  14400       0                0  1.0        0      2  2020  
3257268  14400       0                0  1.0        0      2  2020  
64884    34800       0                0  1.0        0      1  2020


🔍 dot3 duplicate rows:


TRDTYPE  DEPE  COMMODITY2  DISAGMOT  COUNTRY  VALUE  SHIPWT  \
726636        1  0101           7         5     1220   6681       0   
875158        1  0101           7         5     1220   6681       0   
0             1  0101           9         5     1220   2864       0   
66472         1  0101           9         5     1220   2864       0   
149592        1  0101           9         5     1220   2864       0   
382098        1  0101           9         5     1220   2864       0   
513316        1  0101           9         5     1220   2864       0   
695638        1  0101           9         5     1220   2864       0   
794594        1  0101           9         5     1220   2864       0   
165008        1  0101           9         5     1220   8670       0   

        FREIGHT_CHARGES   DF CONTCODE  MONTH  YEAR  
726636              257  1.0        X      9  2020  
875158              257  1.0        X      9  2020  
0                    56  2.0        X      4  2020  
66472                56  2.0        X      4  2020  
149592               56  2.0        X      4  2020  
382098               56  2.0        X      4  2020  
513316               56  2.0        X      4  2020  
695638               56  2.0        X      4  2020  
794594               56  2.0        X      4  2020  
165008              170  2.0        X      5  2020

In [17]:
# Drop the duplicates and reset the index:

dot1 = dot1.drop_duplicates().reset_index(drop=True)
dot2 = dot2.drop_duplicates().reset_index(drop=True)
dot3 = dot3.drop_duplicates().reset_index(drop=True)

# Optional: Check final shape
print("dot1 shape after removing duplicates:", dot1.shape)
print("dot2 shape after removing duplicates:", dot2.shape)
print("dot3 shape after removing duplicates:", dot3.shape)


dot1 shape after removing duplicates: (232029, 14)
dot2 shape after removing duplicates: (634881, 14)
dot3 shape after removing duplicates: (148522, 12)


In [18]:
print(" Data types for dot1:")
print(dot1.dtypes)
print("-" * 40)

print(" Data types for dot2:")
print(dot2.dtypes)
print("-" * 40)

print(" Data types for dot3:")
print(dot3.dtypes)
print("-" * 40)


 Data types for dot1:
TRDTYPE              int64
USASTATE            object
DEPE                object
DISAGMOT             int64
MEXSTATE            object
CANPROV             object
COUNTRY              int64
VALUE                int64
SHIPWT               int64
FREIGHT_CHARGES      int64
DF                 float64
CONTCODE            object
MONTH                int64
YEAR                 int64
dtype: object
----------------------------------------
 Data types for dot2:
TRDTYPE              int64
USASTATE            object
COMMODITY2           int64
DISAGMOT             int64
MEXSTATE            object
CANPROV             object
COUNTRY              int64
VALUE                int64
SHIPWT               int64
FREIGHT_CHARGES      int64
DF                 float64
CONTCODE            object
MONTH                int64
YEAR                 int64
dtype: object
----------------------------------------
 Data types for dot3:
TRDTYPE              int64
DEPE                object
COMMODITY2    